apt-hook/esm-templates.cc

```diff
@@ -0,0 +1,273 @@
+   pkgPolicy *policy = cachefile.GetPolicy();
+
+   if (cache == NULL || policy == NULL)
+      return -1;
+
+   for (pkgCache::PkgIterator pkg = cache->PkgBegin(); !pkg.end(); pkg++)
+   {
+      check_esm_upgrade(pkg, policy, res);
+   }
+   return count;
+}
+
+
+static void process_template_file(
+   std::string template_file_name,
+   std::string static_file_name,
+   std::string esm_a_pkgs_count,
+   std::string esm_a_pkgs,
+   std::string esm_i_pkgs_count,
+   std::string esm_i_pkgs
+) {
+   std::ifstream message_tmpl_file(template_file_name.c_str());
+   if (message_tmpl_file.is_open()) {
+      // This line loads the whole file contents into a string
+      std::string message_tmpl((std::istreambuf_iterator<char>(message_tmpl_file)), (std::istreambuf_iterator<char>()));
+
+      message_tmpl_file.close();
+
+      // Process all template variables
+      std::array<std::string, 4> tmpl_var_names = {
+         ESM_APPS_PKGS_COUNT_TEMPLATE_VAR,
+         ESM_APPS_PACKAGES_TEMPLATE_VAR,
+         ESM_INFRA_PKGS_COUNT_TEMPLATE_VAR,
+         ESM_INFRA_PACKAGES_TEMPLATE_VAR
+      };
+      std::array<std::string, 4> tmpl_var_vals = {
+         esm_a_pkgs_count,
+         esm_a_pkgs,
+         esm_i_pkgs_count,
+         esm_i_pkgs
+      };
+      for (uint i = 0; i < tmpl_var_names.size(); i++) {
+         size_t pos = message_tmpl.find(tmpl_var_names[i]);
+         if (pos != std::string::npos) {
+            message_tmpl.replace(pos, tmpl_var_names[i].size(), tmpl_var_vals[i]);
+         }
+      }
+
+      std::ofstream message_static_file(static_file_name.c_str());
+      if (message_static_file.is_open()) {
+         message_static_file << message_tmpl;
+         message_static_file.close();
+      }
+   } else {
+      remove(static_file_name.c_str());
+   }
+}
+
+void process_all_templates() {
+   int bytes_written;
+   int length;
+
+   // Iterate over apt cache looking for esm packages
+   result res = {0, 0, std::vector<std::string>(), 0, 0, std::vector<std::string>()};
+   get_update_count(res);
+   if (_error->PendingError())
+   {
+      _error->DumpErrors();
+      return;
+   }
+
+   // Compute all strings necessary to fill in templates
+   std::string space_separated_esm_i_packages = "";
+   if (res.esm_i_packages.size() > 0) {
+      for (uint i = 0; i < res.esm_i_packages.size() - 1; i++) {
+         space_separated_esm_i_packages.append(res.esm_i_packages[i]);
+         space_separated_esm_i_packages.append(" ");
+      }
+      space_separated_esm_i_packages.append(res.esm_i_packages[res.esm_i_packages.size() - 1]);
+   }
+   std::string space_separated_esm_a_packages = "";
+   if (res.esm_a_packages.size() > 0) {
+      for (uint i = 0; i < res.esm_a_packages.size() - 1; i++) {
+         space_separated_esm_a_packages.append(res.esm_a_packages[i]);
+         space_separated_esm_a_packages.append(" ");
+      }
+      space_separated_esm_a_packages.append(res.esm_a_packages[res.esm_a_packages.size() - 1]);
+   }
```

Comment:
all of this used to be in `main` but I moved it into this function because it is all necessary whether the old hook is executed or if this function is called by the json hook